### ISSUE TITLE
TiddlyDesktop is crashing on M1 Macs.

### ISSUE THREAD

**Issue by hbayindir** [Jermolene/TiddlyDesktop#224]
TiddlyDesktop is repeatedly crashing on M1 powered MacBook computers. While I'm not knowledgeable about underpinnings of the TiddlyDesktop, I can do testing and guided poking on the system. I'll also add some crash reports as soon as I can.

Crashes happen while working on a Wiki file, and while trying to reopen the software again.

**Comment by istvank** [comment]
I am happy that finally someone else has the same problem, and I'm even happier that you found the pattern.  Indeed, it crashed this morning ten minutes into a Teams meeting. I have a 2019 core i7, and the problem appeared after the first Big Sur security update. I am able to reopen TiddlyDesktop in most cases, though.

**Comment by hbayindir** [comment]
Thanks for the feedback. I'll update the title of the bug.
I've also found out that sometimes the helper is using too much CPU (%100 or more).

Maybe some of the underlying parts of the code needs updating. Since I've moved from Zim to Tiddly to be able to work on Mac unhampered, this is a very important problem for me.

**Comment by qabulin** [comment]
I'm also having constant crashing immediately after open or shortly thereafter some use.  I _do not_ use Teams.  I regularly use TiddlyDesktop many times each week, and believe I first noticed the instability after I upgraded my mac to Big Sur from Catalina.  Since then, TiddlyDesktop kept crashing shortly after open.  I tried uninstalling, and cleared out the following files:

```sh
rm -rf ~/Library/Caches/TiddlyDesktop
rm -rf ~/Library/Saved\ Application\ State/com.tiddlywiki.savedState/
rm -rf ~/Library/Application\ Support/TiddlyDesktop/
rm -rf ~/Library/Preferences/com.tiddlywiki.plist
```

After reinstallation, it still continues to crash.

---

my specs:
* macOS Big Sur 11.2.1 (20D74)
* MBP 2019 i7
* TW 5.1.23
* TiddlyDesktop 0.0.14

**Comment by Jermolene** [comment]
Thanks @hbayindir @istvank @qabulin 

As a first step towards addressing this I've made a first prerelease of v0.0.15 of TiddlyDesktop which just includes an update to the newest version of nwjs (v0.51.1), and no other functional changes:

https://github.com/Jermolene/TiddlyDesktop/releases/tag/v0.0.15-prerelease.1

I'd be grateful if you're able to test it.

**Comment by hbayindir** [comment]
> Thanks @hbayindir @istvank @qabulin
> 
> As a first step towards addressing this I've made a first prerelease of v0.0.15 of TiddlyDesktop which just includes an update to the newest version of nwjs (v0.51.1), and no other functional changes:
> 
> https://github.com/Jermolene/TiddlyDesktop/releases/tag/v0.0.15-prerelease.1
> 
> I'd be grateful if you're able to test it.

Thanks @Jermolene for the new version. Since I hold all my office notes in Tiddly, and would be migrating a lot of information into it, I'd be testing it very extensively and reporting my feedback here.

I've downloaded it already and would be installing it now.

**Comment by hbayindir** [comment]
I've just tried, and it crashes more reliably now. This is how it happens:

1. Tiddly Wiki (wiki selection screen) opens fine.
2. When I open my Wiki file, it crashes on the spot (see the attached crash log).
3. Moreover, when I delete the following files, I can open my Wiki file once, however when I quit Tiddly, it's not possible to reopen it again (i.e. `GOTO 1`)

```
rm -rf ~/Library/Caches/TiddlyDesktop
rm -rf ~/Library/Saved\ Application\ State/com.tiddlywiki.savedState/
rm -rf ~/Library/Application\ Support/TiddlyDesktop/
rm -rf ~/Library/Preferences/com.tiddlywiki.plist
```

I think this situation points to something, which is not written to disk as Tiddly intended to write. As a result, Tiddly blows up in a rather spectacular fashion when it tries to read these.

### Crash Log

* [Tiddly Wiki 0.15.p1 Crash Report 202102192248.txt](https://github.com/Jermolene/TiddlyDesktop/files/6012642/Tiddly.Wiki.0.15.p1.Crash.Report.202102192248.txt)
* [A short video of the problem](https://user-images.githubusercontent.com/479423/108555707-e68e1000-7306-11eb-9a9a-e68beb819101.mov)

**Comment by Jermolene** [comment]
Thanks for the report @hbayindir that's very helpful. Do you get the same crash for all wikis, including a freshly downloaded https://tiddlywiki.com/empty.html?

It looks like we're still some way from an M1 build of nwjs: https://github.com/nwjs/nw.js/issues/7620

**Comment by hbayindir** [comment]
> Thanks for the report @hbayindir that's very helpful. Do you get the same crash for all wikis, including a freshly downloaded https://tiddlywiki.com/empty.html?

You're very welcome. Unfortunately, yes. In the same fashion. I'm adding the crash log again. Is it possible to get a `debug` log either from `TiddlyDesktop` or `nw.js`? I might be able to do some digging in my free time.

### Crash Log

[Tiddly Wiki 0.15.p1 Crash Report 202102200115.txt](https://github.com/Jermolene/TiddlyDesktop/files/6013274/Tiddly.Wiki.0.15.p1.Crash.Report.202102200115.txt)

**Comment by hbayindir** [comment]
> Thanks for the report @hbayindir that's very helpful. Do you get the same crash for all wikis, including a freshly downloaded https://tiddlywiki.com/empty.html?

You're very welcome. Unfortunately yes. Exactly the same fashion. Opening for the first time works, saving works, but when you exit TiddlyDesktop, you cannot return to the wiki again. It crashes in the same, spectacular fashion. The crash log is 
[here](https://github.com/Jermolene/TiddlyDesktop/files/6013300/Tiddly.Wiki.0.15.p1.Crash.Report.202102200115.txt).

Moreover, I've tested the latest version and my "Office Wiki" on an Intel Mac (Intel, Big Sur), and it works absolutely fine. So neither TiddlyDesktop, nor the file has a problem it seems. It's may about the stored state on the disk.  

Is it possible to get debug logging from either `TiddlyDesktop` or `nw.js`? I might try to dig into it if I have free time.

Thanks for your time and quick reply again.

**Comment by nderambure** [comment]
Same problem here with 0.0.14 but it crashes by itself, no need to have m$ teams :)
(On Apple M1 and Big Sur 11.2).

**Comment by mrmattson** [comment]
Same problem here without Microsoft Teams. My specs are similar to qabulin, but with a MBP 2018 i7. Started after upgrading to Big Sur.

**Comment by hbayindir** [comment]
> Same problem here without Microsoft Teams. My specs are similar to qabulin, but with a MBP 2018 i7. Started after upgrading to Big Sur.

@mrmattson, can you please clean the following files and retry using TiddlyWiki? If it's still crashing, I'd like to update the title of the bug accordingly. I don't want to upgrade it twice, back to back.

```
rm -rf ~/Library/Caches/TiddlyDesktop
rm -rf ~/Library/Saved\ Application\ State/com.tiddlywiki.savedState/
rm -rf ~/Library/Application\ Support/TiddlyDesktop/
rm -rf ~/Library/Preferences/com.tiddlywiki.plist
```

**Comment by qabulin** [comment]
> Thanks @hbayindir @istvank @qabulin
> 
> As a first step towards addressing this I've made a first prerelease of v0.0.15 of TiddlyDesktop which just includes an update to the newest version of nwjs (v0.51.1), and no other functional changes:
> 
> https://github.com/Jermolene/TiddlyDesktop/releases/tag/v0.0.15-prerelease.1
> 
> I'd be grateful if you're able to test it.

So far, I've had some success with this today.  I've been successfully running v0.0.15 all day without any crashes.

This is what I did:

1. Removed TiddlyDesktop app.
2. Cleared out save data, cache, etc:
```bash
rm -rf ~/Library/Caches/TiddlyDesktop
rm -rf ~/Library/Saved\ Application\ State/com.tiddlywiki.savedState/
rm -rf ~/Library/Application\ Support/TiddlyDesktop/
rm -rf ~/Library/Preferences/com.tiddlywiki.plist
```
3. Download and installed the v0.0.15 prerelease.
4. Fired up TiddlyDesktop and *changed nothing*.  I kept everything default, out of the box.
5. I selected [Create new wiki] and an "Empty TiddlyWiki file (Prerelease)" to test.
6. I used this for about an hour so, creating fresh Tiddlers, carrying on with my work as normal as I could. (I my existing TiddlyWiki as read-only via Safari whenever I needed to reference old notes).  I tried stopping starting, force quitting, shutting down and starting back up... no crashes.
7. Later, I began customizing by updating settings to my preferred backup location.
8. Then I selected [Add wiki file], choosing my normal existing TiddlyWiki file I've been using almost daily for years up until now.
9. I used my normal wiki for the rest of the day, copying over earlier notes, creating other new notes, editing existing Tiddlers, opening tons of things, normal stuff.  Tried stopping, starting, quitting various ways, shutting down and starting back up... no crashes (so far).

I'll post back if I have any issues.

**Comment by nderambure** [comment]
I can confirm the method used by @qabulin is working for me too. Still on M1 with many apps running, no crash in the day.

**Comment by qabulin** [comment]
@Jermolene, thanks for taking the time to make that prerelease!
Things are looking good right now for me, as I've had no crashes or other noticeable stability issues after replacing TiddlyDesktop with [your prerelease](https://github.com/Jermolene/TiddlyDesktop/issues/224#issuecomment-782272019) and using [the steps I posted earlier](https://github.com/Jermolene/TiddlyDesktop/issues/224#issuecomment-784174127).  I hope others are also doing well! ^_^

**Comment by istvank** [comment]
Thanks for the prerelease! I confirm I had no crashes while working with it during the day, but I MIGHT have some issues after returning from standby. Yesterday after lunch break and standby of my Mac, Tiddly was gone, same this morning, and in both cases I do not remember closing it (I rarely do, in fact). I will continue looking into it. I do not find any crash reports in Library/Application Support (as I did with the previous version MS Teams crash).

**Comment by hbayindir** [comment]
> @Jermolene, thanks for taking the time to make that prerelease!
> Things are looking good right now for me, as I've had no crashes or other noticeable stability issues after replacing TiddlyDesktop with your prerelease and using [the steps I posted earlier](https://github.com/Jermolene/TiddlyDesktop/issues/224#issuecomment-784174127). I hope others are also doing well! ^_^

Did you ever exited and re-opened TiddlyDesktop?
Mine was started to crash after closing and returning to it.

**Comment by qabulin** [comment]
> > @Jermolene, thanks for taking the time to make that prerelease!
> > Things are looking good right now for me, as I've had no crashes or other noticeable stability issues after replacing TiddlyDesktop with your prerelease and using [the steps I posted earlier](https://github.com/Jermolene/TiddlyDesktop/issues/224#issuecomment-784174127). I hope others are also doing well! ^_^
> 
> Did you ever exit and re-open TiddlyDesktop?
> Mine was started to crash after closing and returning to it.
> 
> If yours work without problems, I need to look into mine again.

@hbayindir I believe so.  Please note my steps above:
https://github.com/Jermolene/TiddlyDesktop/issues/224#issuecomment-784174127

Note Step 9:

> [...] Tried stopping, starting, quitting various ways, shutting down and starting back up... no crashes (so far).

Let me know if you have something else in mind for stopping, starting, or quitting/exiting TiddlyDesktop that I can try.  

Otherwise, after following those steps I posted above, I've been using TiddlyDesktop consistently and daily, with no crashing or other unstable behaviors.  It's working just fine for me now. :^)

**Comment by hbayindir** [comment]
@qabulin, thanks for your reply. I've found out why your instructions doesn't help me after the first exit. You're using an intel Mac, mine is an M1. Intel Macs do not crash with latest TiddlyDesktop, you're correct.

My TiddlyDesktop is reliably crashing after exiting, reopening the app and the wiki. Pressing open on my wiki consistently crashes TiddlyDesktop after first exit.

Moreover, I've narrowed the problem. The problematic files are in `~/Library/Application Support/Tiddly Desktop`.  I'll try to further narrow the problem to the file(s) causing the problem. Hope it helps both TiddlyDesktop and nwjs.

Cheers!

**Comment by nderambure** [comment]
Thx for your digging @hbayindir, I'm also on M1 (MacBook Pro), but I can't reproduce your steps, TiddlyDesktop is working fine since the complete wipe and install of 15th version. Please let me know if I can bring more information from my setup.

**Comment by hbayindir** [comment]
> Thx for your digging @hbayindir, I'm also on M1 (MacBook Pro), but I can't reproduce your steps, TiddlyDesktop is working fine since the complete wipe and install of 15th version. Please let me know if I can bring more information from my setup.

You're welcome. I'm just trying to do my best.

That's strange. The behavior I see is as follows:
- After clearing the folder, everything works without any problems.
- When I exit TiddlyDesktop, I cannot re-open any wikis anymore.

Some more observations:
- Newly created wikis open for the first time regardless of state.
    - They cannot be opened after exiting TiddlyDesktop

Can you please post a directory listing of your `~/Library/Application Support/TiddlyDesktop` folder after exiting the application? and `ls -l` will do. You can redact usernames or other sensitive stuff. I just want to see which files remain behind.

**Comment by mrmattson** [comment]
@hbayindir Following up for completeness of data. I accomplished the steps you asked:

```
rm -rf ~/Library/Caches/TiddlyDesktop
rm -rf ~/Library/Saved\ Application\ State/com.tiddlywiki.savedState/
rm -rf ~/Library/Application\ Support/TiddlyDesktop/
rm -rf ~/Library/Preferences/com.tiddlywiki.plist
```

I've had no crashing since doing that 3 days ago. Again, I'm running:

-     macOS Big Sur 11.2.2
-     MBP 2018 i7
-     TW 5.1.23
-     TiddlyDesktop 0.0.14
-     TiddlyWiki file location: Dropbox folder

**Comment by mrmattson** [comment]
@hbayindir Spoke too soon. TiddlyDesktop crashed on me this morning mid-editing a tiddler. I'd just restarted my computer, so I had almost no other applications open. I'd also just opened the TiddlyDesktop app; it had not been running all night. Since 2 days ago I've been opening and closing TiddlyDesktop and making edits to a tiddlywiki file. I'm able to reopen and continue working on TiddlyDesktop now, but I'm no longer crash-free since `rm -rf` those directories.

**Comment by hbayindir** [comment]
Hi @mrmattson, thanks for the information.

It seems that the problem is rooted in how nwjs is handling browser state and profile data it holds at the background, under `~/Library/Application Support/TiddlyDesktop/Default` folder. When I remove it, I'm able to open my wikis again. However, there's something flaky inside because after removing it three or four times (I was trying to capture a screen recording about current state), it stopped crashing, strange. I'll try to dig deeper when I have time.

**TL;DR:** There's a problem about how data is handled under `~/Library/Application Support/TiddlyDesktop/Default` directory. Something in that folder is causing the crash.

**Comment by hbayindir** [comment]
I've probably found the bug. It happens when you exit TiddlyDesktop while a wiki is in fullscreen mode and living on its own dedicated desktop. If you have a windowed Wiki, exiting and restarting doesn't kill anything.

### How to Reproduce
**Prerequisites:** TiddlyDesktop 0.15 prerelease, an M1 powered Mac, macOS Big Sur.

1. First clean problematic files: `rm -rvf ~/Library/Application Support/TiddlyDesktop/Default`.
2. Open TiddlyDesktop.
3. Add your Wiki file.
4. Open your Just added Wiki.
5. Switch to fullscreen (via green button).
6. Exit TiddlyDesktop with `cmd+q`.
7. Re-open TiddlyDesktop.
8. Try to re-open your Wiki.
9. Bingo. Window disappears, you're greeted with a nice crash log.

### Expected Behavior

1. First clean problematic files: `rm -rvf ~/Library/Application Support/TiddlyDesktop/Default`.
2. Open TiddlyDesktop.
3. Add your Wiki file.
4. Open your Wiki.
5. Switch to fullscreen (via green button).
6. Exit TiddlyDesktop with `cmd+q`.
7. Re-open TiddlyDesktop.
8. Try to re-open your Wiki.
9. Your Wiki shall restore. In fullscreen or windowed mode. Doesn't matter.

### Temporary Workaround

Do not exit from TiddlyDesktop while one or more of your wikis are in fullscreen mode.

**Comment by buptsse** [comment]
> > Thanks @hbayindir @istvank @qabulin
> > As a first step towards addressing this I've made a first prerelease of v0.0.15 of TiddlyDesktop which just includes an update to the newest version of nwjs (v0.51.1), and no other functional changes:
> > https://github.com/Jermolene/TiddlyDesktop/releases/tag/v0.0.15-prerelease.1
> > I'd be grateful if you're able to test it.
> 
> So far, I've had some success with this today. I've been successfully running v0.0.15 all day without any crashes.
> 
> This is what I did:
> 
> 1. Removed TiddlyDesktop app.
> 2. Cleared out save data, cache, etc:
> 
> ```shell
> rm -rf ~/Library/Caches/TiddlyDesktop
> rm -rf ~/Library/Saved\ Application\ State/com.tiddlywiki.savedState/
> rm -rf ~/Library/Application\ Support/TiddlyDesktop/
> rm -rf ~/Library/Preferences/com.tiddlywiki.plist
> ```
> 
> 1. Download and installed the v0.0.15 prerelease.
> 2. Fired up TiddlyDesktop and _changed nothing_.  I kept everything default, out of the box.
> 3. I selected [Create new wiki] and an "Empty TiddlyWiki file (Prerelease)" to test.
> 4. I used this for about an hour so, creating fresh Tiddlers, carrying on with my work as normal as I could. (I opened my existing TiddlyWiki as read-only via Safari whenever I needed to reference old notes).  I tried stopping starting, force quitting, shutting down and starting back up... no crashes.
> 5. Later, I began customizing by updating settings to my preferred backup location on my local machine.
> 6. Then I selected [Add wiki file], choosing my normal existing TiddlyWiki file I've been using almost daily for years up until now, which sits in an iCloud Drive folder.
> 7. I used my normal wiki for the rest of the day, copying over earlier notes, creating other new notes, editing existing Tiddlers, opening tons of things, normal stuff.  Tried stopping, starting, quitting various ways, shutting down and starting back up... no crashes (so far).
> 
> I'll post back if I have any issues.
> 
> _(Note: I also edited my earlier provided specs to include my file locations, in case that is ever pertinent.)_

work fine for me! thanks!

**Comment by replaid** [comment]
I'm on 

- Big Sur 11.2.3
- 2017 MBP i5
- TiddlyDesktop 0.0.14
- TiddlyWiki 5.1.23
- frequently have two 4K monitors attached as well as laptop monitor

I never full-screen TiddlyDesktop. I have been wondering why it seems to disappear periodically, and just found it had segfaulted. 

```
com.apple.xpc.launchd[1] (application.com.tiddlywiki.3648420.3648423[53363]): Service exited due to SIGSEGV | sent by exc handler[53363]
```

I am not seeing crash reports in Console.app, only that line. Happened 2x today 6 hours apart and have noticed this other days. I think it's happening while TiddlyDesktop is idle (but not necessarily the machine, this has definitely happened while I'm using other apps), have not had it happen during active use of TiddlyDesktop. 

Please advise whether this is the right issue, happy to help any way I can. I have now started by running the pre-release version.

**Comment by qabulin** [comment]
Oh no. I'm crashing too, but I think I have a workaround.
I'm unsure if I tested out the fullscreen properly before, but everything else has been perfectly stable until just now.  Today, I decided to try fullscreen exit and restart after @replaid posted.

**steps to reproduce:**

1. open TiddlyDesktop
1. open a TiddlyWiki
1. go fullscreen with that TiddlyWiki window
1. Cmd+Q close
1. open TiddlyDesktop app again
1. try to open the previous TiddlyWiki again, and everything (TiddlyDesktop & TiddlyWiki) silently close

I captured a stack trace by running the contents of the application directly from Finder (details expandable below) and repeating the steps above.
<details>
  <summary>click to reveal terminal stack trace...</summary>

```
% /Applications/TiddlyDesktop.app/Contents/MacOS/nwjs ; exit;
Received signal 11 SEGV_MAPERR 000000000030
 [0x00010afba6d9]
 [0x00010aef1393]
 [0x00010afba5f1]
 [0x7fff2038dd7d]
 [0x7fbf4d606dc0]
 [0x00010e9440e6]
 [0x00010eb276f1]
 [0x7fff2043721d]
 [0x7fff204d2c7c]
 [0x7fff204d2bfa]
 [0x7fff2040829e]
 [0x7fff211732c8]
 [0x7fff23764657]
 [0x7fff23146e28]
 [0x7fff235b69c7]
 [0x7fff23148540]
 [0x7fff2371a565]
 [0x00010e8158d0]
 [0x00010eb20892]
 [0x00010eb30ede]
 [0x00010e915646]
 [0x00010e914da5]
 [0x00010e3ba833]
 [0x000109120267]
 [0x0001091230b5]
 [0x00010912370b]
 [0x00010914e9c8]
 [0x00010914e845]
 [0x00010e3ea8fd]
 [0x000108da0dc3]
 [0x000108bd8765]
 [0x00010c123435]
 [0x00010af64b66]
 [0x00010af7b586]
 [0x00010af7b27a]
 [0x00010afd22a3]
 [0x00010afcab92]
 [0x00010afd1c6f]
 [0x7fff20441a8c]
 [0x7fff204419f4]
 [0x7fff20441774]
 [0x7fff2044019c]
 [0x7fff2043f75c]
 [0x7fff28661203]
 [0x7fff28660f65]
 [0x7fff28660d03]
 [0x7fff22c3bb32]
 [0x7fff22c3a305]
 [0x00010b5d6cb0]
 [0x00010afcab92]
 [0x00010b5d6be9]
 [0x7fff22c2c679]
 [0x00010afd2d5c]
 [0x00010afd1793]
 [0x00010af7be2b]
 [0x00010af4333d]
 [0x00010b1ee223]
 [0x000108830635]
 [0x000108832482]
 [0x00010882d860]
 [0x00010ad0cb12]
 [0x00010ad0c723]
 [0x00010ad0ab89]
 [0x00010ad0b3fc]
 [0x0001070bb787]
 [0x000101d1795f]
 [0x7fff20363f3d]
[end of stack trace]
[0602/075314.990303:WARNING:process_memory_mac.cc(93)] mach_vm_read(0x7ffeedeea000, 0x2000): (os/kern) invalid address (1)
[0602/075315.151011:WARNING:crash_report_exception_handler.cc(240)] UniversalExceptionRaise: (os/kern) failure (5)
zsh: segmentation fault  /Applications/TiddlyDesktop.app/Contents/MacOS/nwjs
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
Deleting expired sessions...       5 completed.

[Process completed]
```

</details>

The result is I can no longer access that TiddlyWiki as it will just silently close out everything.  That is, unless I clear the applications window preference cache.

**workaround steps to restore a TiddlyWiki that will no longer open after fullscreen exit:**

1. completely close out TiddlyDestop
1. **CAUTION: This will change the preferences of all Tiddly windows**<br>move (or delete) the Application Support Preferences application cache file from Finder or from the terminal<br>`mv ~/Library/Application\ Support/TiddlyDesktop/Default/Preferences{,.bak}`
1. start TiddlyDesktop, which will generate a new Preferences file
1. open TiddlyWiki again and note window caching has been reset

alternatives:

A less impactful (does not reset all Tiddly window cache) but possibly risky fix (if you fat-finger something) would be to open the Preference JSON file, and look for `"fullscreen":true` and change it to `false`.  There will be sections called `wikifile:///`... and you should see your wiki's location, and under the "html" are the cached window preferences.  **Use at your own risk**, but I made a lazy fix-it script.  It may blindly smash things but then I can just delete the entire preferences file and start fresh if it wrecks my preferences file:

```bash
#!/bin/bash
# USE AT YOUR OWN RISK
PREF_FILE=${HOME}/Library/Application\ Support/TiddlyDesktop/Default/Preferences
sed -i .bak -e 's/\"fullscreen\":true/\"fullscreen\":false/g' "$PREF_FILE" 
```

---

my specs:

- macOS Big Sur 11.3.1 (20E241)
- MBP 2019 i7
- TW 5.1.23
- TiddlyDesktop 0.0.14 (15) (a [prerelease](https://github.com/Jermolene/TiddlyDesktop/issues/224#issuecomment-782272019) of v0.0.15)
- backup location: local disk
- TiddlyWiki file location: iCloud Drive folder

**Comment by qabulin** [comment]
Oh, I just discovered another issue:
Each time you fullscreen crash TiddlyDesktop, there is a left over "nwjs Helper (Renderer)" process consuming an appreciable amount of cpu.  After testing the above, my CPU fan was running high (b/c temp at 88.05 C), and my system was running slower.  I found this:

```
% pgrep nwjs
73020
73065
73087
73117
73140
73167
73219
73239
73261
```

Each of those was consuming about 30-50% CPU.  I had **no other nwjs apps running**, so I just killed them after closing TiddlyDesktop:

```
% pgrep nwjs | xargs kill && pgrep nwjs

%
```

Now everything is back to normal.  (I may add this to my script.)

**Comment by replaid** [comment]
I made no other changes at all, simply started running the 0.0.15 prerelease. I started using my computer and found TiddlyWiki slow to respond. Noticed my CPU fan running, popped open Activity Monitor and saw this at the top.

![Screen Shot 2021-06-03 at 12 07 10](https://user-images.githubusercontent.com/52578/120619217-72661b00-c464-11eb-9e32-100dcf87778c.png)

Interestingly, perhaps putting more detail into the idea of this being a "fullscreen" issue, my machine struggles to get connected to both of my 4K monitors, and today it shifted TiddlyWiki from one of the big screens to the main laptop screen. So there could be some buggy display-handling logic that isn't necessarily specific to fullscreen mode (which I have not been using at all lately, never with TiddlyDesktop).

When I hit the Sample button in the nwjs Helper (Renderer) process's window, all the parts of the sample seem to be completely blank. When I try to save the sample, it prompts me for a place to save it, but nothing gets saved. 

<details>
<summmary>So in lieu of that, here's the information in the window that pops up when I double-click on that process's line in the list.</summary>
![Screen Shot 2021-06-03 at 12 24 11](https://user-images.githubusercontent.com/52578/120621551-a2aeb900-c466-11eb-8dff-5d3b61631de1.png)
![Screen Shot 2021-06-03 at 12 24 43](https://user-images.githubusercontent.com/52578/120621632-b9551000-c466-11eb-9777-33a3ac3bcb00.png)

Open Files and Ports:
```
cwd
cwd|rtd info error: No such file or directory
txt
/Library/Preferences/Logging/.plist-cache.C64Lj3Vs
txt
/System/Library/Fonts/SFNSItalic.ttf
txt
/usr/lib/dyld
txt
/private/var/db/timezone/tz/2021a.1.0/icutz/icutz44l.dat
txt
/usr/share/icu/icudt66l.dat
txt
/System/Library/Fonts/SFNS.ttf
txt
/System/Library/Fonts/Helvetica.ttc
txt
/System/Library/Fonts/Supplemental/Arial.ttf
txt
/System/Library/Fonts/Apple Color Emoji.ttc
txt
/System/Library/Fonts/Times.ttc
txt
/System/Library/Fonts/Menlo.ttc
0
/dev/null
1
/dev/null
2
/dev/null
3
(revoked)
4
(revoked)
5
(revoked)
6
(revoked)
7
(revoked)
8
(revoked)
9
->0x1a8a71a110891a16
10
->0x635d64362e931fd
11
count=0, state=0xb
12
->0xefea0d30f9cc1131
13
->0x4c98e2d18c34dbe
14
->0x483e207512bac5ce
15
->0x8482c5667df8a9fd
16
->0x8478500772d87fae
17
->0xd7ec441a7c3feae5
18
count=0, state=0x12
19
count=0, state=0x12
20
/dev/urandom
21
/private/var/folders/jr/7_yfzsds0k5fsxj3trm81mwh0000gp/T/.com.tiddlywiki.6pYE4P
22
count=0, state=0xa
23
->0xf823a7039295cb61
24
->0x4e9f9653e5db61ee
25
->0x9df3c9507ad08291
26
->0x81a1ec2f0d9de30
27
count=0, state=0xa
28
->0x9ee208d1a0389726
29
->0xe81f8a19b7930b0d
30
->0xd0f82cd62efb38f0
31
->0xbcdb775c3e49dd34
32
count=0, state=0xa
33
->0x2608543607dcad02
34
->0xf62cc98510110094
35
->0x16d3c77c66899d43
36
->0xed550b117170a5de
37
count=0, state=0xa
38
->0xb94af2f67def7a3f
39
->0xcdf9483a5f4aaba
40
->0xc95c474ca2e23e67
41
->0x7191d8fff6a71264
```

</details>

<details>
<summary>
Interestingly, there's also a `chrome_crashpad_handler` taking 80% of CPU. Also blank when I try to sample it.
</summary>

![Screen Shot 2021-06-03 at 12 34 52](https://user-images.githubusercontent.com/52578/120623154-24ebad00-c468-11eb-9e23-70c5e283e88e.png)

![Screen Shot 2021-06-03 at 12 35 29](https://user-images.githubusercontent.com/52578/120623220-36cd5000-c468-11eb-95c5-065c483e244b.png)

Open Files and Ports:

```
cwd
/
txt
/Library/Preferences/Logging/.plist-cache.C64Lj3Vs
txt
/Users/john/Library/Application Support/TiddlyDesktop/CrashpadMetrics-active.pma
txt
/usr/lib/dyld
0
/dev/null
1
/dev/null
2
/dev/null
3
/Users/john/Library/Application Support/TiddlyDesktop/CrashpadMetrics-active.pma
```

`/Library/Preferences/Logging/.plist-cache.C64Lj3Vs` is an empty file.

Let me know if the contents of `~/Library/Application Support/TiddlyDesktop/Crashpad/completed/*.dmp` would be helpful to have. [(NW.js documentation on crash dumps.)](https://nwjs.readthedocs.io/en/latest/For%20Developers/Understanding%20Crash%20Dump/)

</details>

Seems like it's worth dropping a link to https://github.com/nwjs/nw.js/issues/7684 here.

**Comment by hbayindir** [comment]
Since I stopped using TiddlyWiki in fullscreen mode in my MacBook Air, it didn't create any problems for me, but it's true that nw.js has some work to do to fix these problems.

Unfortunately it seems nw.js both very popular and somewhat understaffed and swamped with requests. I'm not sure that I can debug and write fixes for it, but we may need to dig the problem deeper and help the developers somewhat.

Any ideas?

**Comment by qabulin** [comment]
It seems like the nightly releases may be making some progress in reducing the crashing, per https://github.com/nwjs/nw.js/issues/7697 . I'm curious if anyone may have the bandwidth to try this (unfortunately I'm not free at the moment).

**Comment by Jermolene** [comment]
Thanks for your patience everyone. There's a new prerelease here if you are able to give it a test:

https://github.com/TiddlyWiki/TiddlyDesktop/releases/tag/v0.0.15-prerelease.4

**Comment by rhysnewell** [comment]
I believe I am still experience this same issue with the latest pre-release. Apple M1, creating a new wiki is fine but closing and re-opening the wiki will cause an instant crash

**Comment by S4GU4R0** [comment]
As long as I don't open classic wikis while in fullscreen mode, I can avoid the app crashing. When I forget and walk right into the crash, there are alerts regarding nw.js crashing first. Not sure how helpful that is but leaving it here just in case.
